packages/components/src/components/PipelineRun/PipelineRun.js

```diff
@@ -248,7 +249,46 @@ export /* istanbul ignore next */ class PipelineRunContainer extends Component {
         </>
       );
     }
-    const taskRuns = this.loadTaskRuns(pipelineRun, taskRunNames);
+    let taskRuns = this.loadTaskRuns(pipelineRun, taskRunNames);
+
+    function findEarliestStepForTask(steps) {
+      return steps[0];
+    }
+
+    taskRuns = taskRuns.sort(function(taskA, taskB) {
+      if (taskA && taskB) {
+        const foundEarliestTaskAStep = findEarliestStepForTask(taskA.steps);
+        const foundEarliestTaskBStep = findEarliestStepForTask(taskB.steps);
+        if (foundEarliestTaskAStep && foundEarliestTaskBStep) {
+          if (
+            foundEarliestTaskAStep.terminated &&
+            foundEarliestTaskBStep.terminated
+          ) {
+            const startedAtA = new Date(
+              foundEarliestTaskAStep.terminated.startedAt
+            ).getTime();
+            const startedAtB = new Date(
+              foundEarliestTaskBStep.terminated.startedAt
+            ).getTime();
+            return startedAtA - startedAtB;
+          }
+          if (
+            foundEarliestTaskAStep.running &&
+            foundEarliestTaskBStep.running
+          ) {
+            const startedAtA = new Date(
+              foundEarliestTaskAStep.running.startedAt
+            ).getTime();
+            const startedAtB = new Date(
+              foundEarliestTaskBStep.running.startedAt
+            ).getTime();
+            return startedAtA - startedAtB;
+          }
+        }
+        return 0;
+      }
+      return 0;
+    });
```

Comment:
Can we extract this to a function to keep `render` a bit more manageable? It can live in the same file, outside the component.

Reply:
I'm also not 100% sure when it should sort vs. leaving the order unchanged. The code here will only sort them when both steps are in the same state (i.e. both terminated, or both running). It leaves them as they are (`return 0`) in the case that either of them are in the 'waiting' state, or if they're in different states. Is this desired?

Early returns could be used to reduce some of the indentation here and (IMO) make it easier to read.

Depending on exactly when we want to sort the tasks, something like this might work:

```js
taskRuns = taskRuns.sort((taskRunA, taskRunB) => {
  if (!taskRunA || !taskRunB) {
    return 0;
  }

  const firstStepA = taskRunA.steps[0];
  const firstStepB = taskRunB.steps[0];

  if (!firstStepA || !firstStepB) {
    // can we actually reach this state?
    return 0;
  }

  const { startedAt: startedAtA } = firstStepA.terminated || firstStepA.running || {};
  const { startedAt: startedAtB } = firstStepB.terminated || firstStepB.running || {};

  if (!startedAtA || !startedAtB) {
    return 0;
  }

  return new Date(startedAtA).getTime() - new Date(startedAtB).getTime();
});
```

Reply:
Yeah this looks so much neater, thanks, so for

```
The code here will only sort them when both steps are in the same state (i.e. both terminated, or both running). It leaves them as they are (return 0) in the case that either of them are in the 'waiting' state, or if they're in different states. Is this desired?
```

you're right, I'm willing to live with this though - and the sorting happens often, on clicking the tasks or steps for example. I figured this would cause undesired behaviour (e.g. clicking on what you think is Step B might end up being Step A, but actually things are fine), I've addressed your other comments too 👀